conan_provider.cmake

```diff
@@ -1,7 +1,26 @@
 set(CONAN_MINIMUM_VERSION 2.0.5)
 
 
-function(detect_os OS OS_VERSION)
+function(detect_arch ARCH)
```

Comment:
This function should probably get the already detected OS as an input argument to decide based on it? 
In the case of ANDROID is more evident, but what if we need to check arch for others like QNX or Mac, which OS detection already contains some logic?

Reply:
That's definitely a good thing to decide on and be consistent with. Do we use `OS` or do we use the global CMake platform variables such as `ANDROID`, `IOS`, `MSVC`, `APPLE`, etc..? I would prefer the CMake platform variables over `OS` for the following reasons:

- Using a variable named `OS` outside of the context of setting the Conan profile's `os` setting might be confusing in cross-compiling scenarios as it might be unclear whether it's referring to the underlying build platform or the target platform
- We're already heavily using global CMake variables to build the Conan profile so why not use the platform variables too?
- `APPLE` would be convenient for macOS and iOS platforms since its architecture settings for both platforms are set in the same CMake variable

I don't have a strong preference though so I'm happy to change it and continue on that pattern for future pull requests.

Reply:
I don't have a strong preference either. I was mostly thinking that at the moment ``elseif(${CMAKE_SYSTEM_NAME} STREQUAL "QNX")`` and ``if(${CMAKE_SYSTEM_NAME} STREQUAL "Darwin")``, that is, based on CMAKE_SYSTEM_NAME instead of the CMake variables. 

I am not opposed to change the detection of OS, to make it consistent, but the same ``ifs`` should be applied in the different places. Maybe change that first in a previous PR?

Reply:
Sure, I'll try that out, probably not until the weekend, it's a busy week.

Reply:
After thinking more about it, I think we should stick with `CMAKE_SYSTEM_NAME`. The simplicity of setting its value as the value for the `os` Conan setting is going to work for a lot of platforms, including Android and iOS and even exotic platforms with custom toolchains might 'just work'.

So I'll take the route in your original comment:
- All `if(ANDROID)` get changed to `if(${CMAKE_SYSTEM_NAME} STREQUAL "Android")`
- `OS` variable is passed to `detect_arch` and `detect_lib_cxx` and `if(${OS} STREQUAL "Android")` can be used for any Android-specific logic.

Sound good?

Reply:
Yes, exactly!